lib/autoproj/cli/envsh.rb

```diff
@@ -1,19 +1,147 @@
+require 'rbconfig'
 require 'autoproj/cli/inspection_tool'
 module Autoproj
     module CLI
         class Envsh < InspectionTool
-            def validate_options(_unused, options = Hash.new)
-                _, options = super(_unused, options)
+            attr_predicate :watch?
+            attr_reader :options
+            attr_reader :source_packages
+            attr_reader :notifier
+            attr_reader :watchers
+            def validate_options(unused, options = {})
+                _, options = super(unused, options)
+                @watchers = []
+                @watch = options[:watch]
+                @options = options
                 options
             end
 
-            def run(**options)
+            def update_workspace(restart = false)
+                exec('autoproj', 'envsh', '--watch') if restart
```

Comment:
This method is now doing two completely unrelated things based on the restart flag. Separate the methods into `update_workspace` and `restart` (and *definitely* split envsh from watch).